## Guided Project: Building Fast Queries on a CSV
## Reading the Inventory
Use the csv module to read the laptops.csv file and separate the header from the rows.

In [10]:
import csv
with open ('laptops.csv') as file:
    reader = csv.reader(file)
    rows = list(reader)
    header = rows[0]
    rows = rows[1:]
print(header)
print(rows[:5])

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']
[['6571244', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', '1339'], ['7287764', 'Apple', 'Macbook Air', 'Ultrabook', '13.3', '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', '898'], ['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', '575'], ['9722156', 'Apple', 'MacBook Pro', 'Ultrabook', '15.4', 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', '2537'], ['8550527', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 3.1GHz', '8GB', '256GB 

## Inventory Class
Start implementing a class to represent the inventory. It get the name of the CSV file as argument and reads it into self.header and self.rows.

In [11]:
class Inventory():
    def __init__(self, csv_filename):
        with open ('laptops.csv') as file:
            reader = csv.reader(file)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1] = int(row[-1])

inventory = Inventory('laptops.csv')
print(inventory.header)
print(len(inventory.rows))

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']
1303


## Finding a Laptop From the Id
Implement a get_laptop_from_id() function that given a laptop identifier find the row corresponding to that laptop.

In [12]:
class Inventory():
    def __init__(self, csv_filename):
        with open ('laptops.csv') as file:
            reader = csv.reader(file)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
inventory = Inventory('laptops.csv')
print(inventory.get_laptop_from_id('3362737'))
print(inventory.get_laptop_from_id('3362736'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]
None


## Improving Id Lookups
Improve the time complexity of finding a laptop with a given id by precomputing a dictionary that maps laptop ids to rows.

In [13]:
class Inventory():
    def __init__(self, csv_filename):
        with open ('laptops.csv') as file:
            reader = csv.reader(file)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
            
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        else:
            return None
        
inventory = Inventory('laptops.csv')
print(inventory.get_laptop_from_id_fast('3362737'))
print(inventory.get_laptop_from_id_fast('3362736'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', '575']
None


## Comparing Performance
Compare the performance of both function for id lookup.

In [24]:
import time
import random

class Inventory():
    def __init__(self, csv_filename):
        with open ('laptops.csv') as file:
            reader = csv.reader(file)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        else:
            return None


ids = [str(random.randint(1000000, 9999999)) for _ in range(10000)]

inventory = Inventory('laptops.csv')
total_time_no_dict = 0

for el in ids:
    start = time.time()
    inventory.get_laptop_from_id(el)
    end = time.time()
    total_time_no_dict += end - start

total_time_dict = 0
for el in ids:
    start = time.time()
    inventory.get_laptop_from_id_fast(el)
    end = time.time()
    total_time_dict += (end - start)

print(total_time_no_dict)
print(total_time_dict)


0.9643886089324951
0.0043179988861083984


<IPython.core.display.Markdown object>

### Analysis

We can see a significant improve in performance. If we divide 0.964 by 0.004 we see that the new method is about 241 times faster for this input size.

## Two Laptop Promotion
Write a method that finds whether we can spend a given amount of money by purchasing either one or two laptops.

In [15]:
class Inventory():
    def __init__(self, csv_filename):
        with open ('laptops.csv') as file:
            reader = csv.reader(file)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        self.id_to_row = {}
        for row in self.rows:              
            row[-1] = int(row[-1])
        for row in self.rows:
            self.id_to_row[row[0]] = row
        
        
            
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        else:
            return None
    
    def check_promotion_dollars(self, dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row2[-1] + row2[-1] == dollars:
                    return True
        return False
inventory = Inventory('laptops.csv')
print(inventory.check_promotion_dollars(1000))
print(inventory.check_promotion_dollars(442))

True
False


## Optimizing Laptop Promotion
Create a faster version of the promotion method by using the techniques we've learned in the course

In [16]:
class Inventory():
    def __init__(self, csv_filename):
        with open ('laptops.csv') as file:
            reader = csv.reader(file)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        self.id_to_row = {}
        self.prices = set()
        for row in self.rows:              
            row[-1] = int(row[-1])
            self.prices.add(row[-1])
        for row in self.rows:
            self.id_to_row[row[0]] = row
        
        
            
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        else:
            return None
    
    def check_promotion_dollars(self, dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row2[-1] + row2[-1] == dollars:
                    return True
        return False
    
    
    def check_promotion_dollars_fast(self, dollars):
        if dollars in self.prices:
            return True
        for price in self.prices:
            if (dollars - price) in self.prices:
                return True
        return False

inventory = Inventory('laptops.csv')
print(inventory.check_promotion_dollars_fast(1000))
print(inventory.check_promotion_dollars_fast(442))

True
False


In [26]:
## Comparing Promotion Functions
Compare the performance of both methods for the promotion.

SyntaxError: invalid syntax (<ipython-input-26-251bddf7229f>, line 2)

In [19]:
import time
import random


class Inventory():
    def __init__(self, csv_filename):
        with open (csv_filename) as file:
            reader = csv.reader(file)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        self.id_to_row = {}
        self.prices = set()
        for row in self.rows:              
            row[-1] = int(row[-1])
            self.prices.add(row[-1])
        for row in self.rows:
            self.id_to_row[row[0]] = row
        
        
            
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        else:
            return None
    
    def check_promotion_dollars(self, dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row2[-1] + row2[-1] == dollars:
                    return True
        return False
    
    
    def check_promotion_dollars_fast(self, dollars):
        if dollars in self.prices:
            return True
        for price in self.prices:
            if (dollars - price) in self.prices:
                return True
        return False

    
prices = [random.randint(100, 5000) for _ in range(100)]
inventory = Inventory('laptops.csv')

total_time_no_set = 0
for el in prices:
    start = time.time()
    inventory.check_promotion_dollars(el)
    end = time.time()
    total_time_no_set += end - start


total_time_set = 0
for el in prices:
    start = time.time()
    inventory.check_promotion_dollars_fast(el)
    end = time.time()
    total_time_set += end - start

print(total_time_no_set)
print(total_time_set)

21.979313135147095
0.0006880760192871094


Analysis

We can see a significant improve in performance. If we divide 21.979 by 0.0006 we see that the new method is about 36,632 times faster for this input size!



In [18]:
def row_price(row):
    return row[-1]

class Inventory():
    def __init__(self, csv_filename):
        with open (csv_filename) as file:
            reader = csv.reader(file)
            rows = list(reader)
        self.header = rows[0]
        self.rows = rows[1:]
        for row in self.rows:              
            row[-1] = int(row[-1])
        self.rows_by_price = sorted(self.rows, key=row_price)
    
    def find_first_laptop_more_expensive(self, target_price):
        range_start = 0
        range_end = len(self.rows_by_price) - 1
        while range_start < range_end:
            range_middle = (range_end + range_start) // 2  
            price = self.rows_by_price[range_middle][-1]
            if price > target_price:
                range_end = range_middle
            else:
                range_start = range_middle + 1
        if self.rows_by_price[range_start][-1] <= target_price:
            return -1
        return range_start


inventory = Inventory('laptops.csv')
print(inventory.find_first_laptop_more_expensive(1000))
print(inventory.find_first_laptop_more_expensive(10000))       

683
-1
